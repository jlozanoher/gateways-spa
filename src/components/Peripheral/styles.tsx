import * as Antd from "antd";
import styled from "styled-components";

export const Space = styled(Antd.Space)`
  gap: 2px !important;
`;

export const Card = styled(Antd.Card)`
  .ant-card-body {
    padding: 8px 4px;
  }
  width: 160px;
`;
