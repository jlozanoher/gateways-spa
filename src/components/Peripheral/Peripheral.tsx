import { DeleteOutlined, EditOutlined } from "@ant-design/icons";
import { Button, Typography } from "antd";
import cs from "classnames";
import dayjs from "dayjs";
import React, { useContext } from "react";
import { useDrag } from "react-dnd";
import { PeripheralContext } from "../../App";
import { usePeripheralDelete } from "../../contollers/peripheral.controller";
import { dotsString } from "../../lib/utils";
import { ItemTypes } from "../../types/ItemsTypes";
import { PeripheralModel } from "../../types/peripheral.model";
import * as S2 from "../styles";
import * as S from "./styles";

const { Text } = Typography;

interface Props {
  peripheral: PeripheralModel;
}

export const Peripheral = (props: Props) => {
  const { peripheral } = props;

  const [{ opacity }, dragRef] = useDrag(
    () => ({
      type: ItemTypes.PERIPHERAL,
      item: peripheral,
      collect: (monitor) => ({
        opacity: monitor.isDragging() ? 0.5 : 1,
      }),
    }),
    []
  );

  const {
    fetchPeripherals = () => {},
    setPeripheralForEdition = () => {},
    setShowModal = () => {},
  } = useContext(PeripheralContext);

  const { handleDeletePeripheral } = usePeripheralDelete({
    onSuccess: () => {
      fetchPeripherals();
    },
  });

  const onDelete = () => {
    handleDeletePeripheral(peripheral._id);
  };

  const onEdit = () => {
    setPeripheralForEdition(peripheral);
    setShowModal(true);
  };

  return (
    <div ref={dragRef} style={{ opacity, width: "180px" }}>
      <S.Card hoverable className={cs({ connected: peripheral.gateway })}>
        <S2.Box>
          <S2.BoxC>
            <Button icon={<DeleteOutlined />} danger onClick={onDelete} />
            <Button icon={<EditOutlined />} onClick={onEdit} />
          </S2.BoxC>

          <S.Space direction="vertical" className={cs("small")}>
            <Text>{peripheral.uid}</Text>
            <Text type="secondary">{dotsString(peripheral.gateway)}</Text>
            <Text>{peripheral.status}</Text>
            <Text>{peripheral.vendor}</Text>
            <Text>{dayjs(peripheral.createdAt).format("YY-MM-DD")}</Text>
          </S.Space>
        </S2.Box>
      </S.Card>
    </div>
  );
};
