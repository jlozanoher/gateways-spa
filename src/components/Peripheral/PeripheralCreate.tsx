import { Button, Form, Input, InputNumber, Select } from "antd";
import dayjs from "dayjs";
import React, { useContext, useEffect } from "react";
import { AppContext } from "../../App";
import {
  usePeripheralCreate,
  usePeripheralUpdate,
} from "../../contollers/peripheral.controller";
import {
  PeripheralModel,
  PeripheralStatusEnum,
} from "../../types/peripheral.model";
import DatePicker from "../DatePicker";
import * as S from "../styles";

const { Option } = Select;

const layout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 16 },
};
const tailLayout = {
  wrapperCol: { offset: 8, span: 16 },
};

interface Props {
  peripheral?: PeripheralModel; // For updating instead of creating
  onSave?: () => void;
}

export const PeripheralCreate = (props: Props) => {
  const { peripheral, onSave = () => {} } = props;
  const { fetchPeripherals = () => {} } = useContext(AppContext);

  const [form] = Form.useForm();

  const onError = (err: any) => {
    if (err?.name?.code === 11000 && err?.name?.keyPattern?.uid) {
      form.setFields([
        {
          name: "uid",
          errors: ["Duplicity"],
        },
      ]);
    } else if (err?.name === "ValidationError") {
      if (err?.params?.path === "body.uid")
        form.setFields([
          {
            name: "uid",
            errors: [err?.message],
          },
        ]);
    }
  };

  useEffect(() => {
    if (peripheral) {
      form.setFieldsValue({
        ...peripheral,
        createdAt: dayjs(peripheral.createdAt),
      });
    }
  }, [peripheral]);

  const { handleCreatePeripheral } = usePeripheralCreate({
    onSuccess: () => {
      fetchPeripherals();
      onSave();
    },
    onError: onError,
  });

  const { handleUpdatePeripheral } = usePeripheralUpdate({
    onSuccess: () => {
      fetchPeripherals();
      onSave();
    },
    onError: onError,
  });

  const onFinish = (values: any) => {
    if (peripheral?._id) {
      handleUpdatePeripheral({ ...values, _id: peripheral._id });
    } else {
      handleCreatePeripheral(values);
    }
  };

  const onClear = () => {
    form.resetFields();
  };

  return (
    <div>
      <Form
        {...layout}
        form={form}
        onFinish={onFinish}
        initialValues={{ status: PeripheralStatusEnum.offline }}
      >
        <Form.Item name={"uid"} label={"uid"} rules={[{ type: "number" }]}>
          <InputNumber
            onChange={(v) => form.setFields([{ name: "uid", errors: [] }])}
          />
        </Form.Item>
        <Form.Item
          name={"vendor"}
          label={"vendor"}
          rules={[{ required: true }]}
        >
          <Input />
        </Form.Item>
        <Form.Item name={"status"} label={"status"}>
          <Select showSearch placeholder="status">
            <Option value={PeripheralStatusEnum.online} key={1}>
              {PeripheralStatusEnum.online}
            </Option>
            <Option value={PeripheralStatusEnum.offline} key={2}>
              {PeripheralStatusEnum.offline}
            </Option>
          </Select>
        </Form.Item>
        <Form.Item name={"createdAt"} label={"createdAt"}>
          <DatePicker />
        </Form.Item>
        <Form.Item {...tailLayout}>
          <S.Box>
            <Button type="primary" htmlType="submit">
              {peripheral?._id ? "Update Peripheral" : "Create Peripheral"}
            </Button>
            <Button onClick={onClear}>Clear</Button>
          </S.Box>
        </Form.Item>
      </Form>
    </div>
  );
};
