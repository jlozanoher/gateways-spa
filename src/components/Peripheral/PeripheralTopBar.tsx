import { PlusOutlined } from "@ant-design/icons";
import { Button, Typography } from "antd";
import React, { useContext } from "react";
import { AppContext } from "../../App";
import * as S from "../styles";

const { Title } = Typography;

export const PeripheralTopBar = () => {
  const { setShowModal = () => {} } = useContext(AppContext);

  const onClick = () => {
    setShowModal(true);
  };

  return (
    <S.Header>
      <Title level={3}>Peripherals</Title>
      <S.Grow />
      <Button type="primary" icon={<PlusOutlined />} onClick={onClick}>
        Add Peripheral
      </Button>
    </S.Header>
  );
};
