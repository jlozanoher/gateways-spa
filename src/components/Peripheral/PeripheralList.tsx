import { List } from "antd";
import React from "react";
import { PeripheralModel } from "../../types/peripheral.model";
import { Peripheral } from "./Peripheral";

interface Props {
  peripherals: PeripheralModel[];
}

export const PeripheralList = (props: Props) => {
  const { peripherals } = props;
  return (
    <List
      grid={{
        gutter: 8,
        xs: 1,
        sm: 2,
        md: 4,
        lg: 4,
        xl: 6,
        xxl: 3,
      }}
      dataSource={peripherals}
      renderItem={(peripheral) => (
        <List.Item>
          <Peripheral key={peripheral._id} peripheral={peripheral} />
        </List.Item>
      )}
    />
  );
};
