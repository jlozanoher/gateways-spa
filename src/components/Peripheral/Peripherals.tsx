import { Button } from "antd";
import React, { useContext } from "react";
import { useDrop } from "react-dnd";
import { PeripheralContext } from "../../App";
import { usePeripheralUpdate } from "../../contollers/peripheral.controller";
import { ItemTypes } from "../../types/ItemsTypes";
import * as S from "../styles";
import { PeripheralCreate } from "./PeripheralCreate";
import { PeripheralList } from "./PeripheralList";
import { PeripheralTopBar } from "./PeripheralTopBar";

export const Peripherals = () => {
  const {
    peripherals = [],
    showModal,
    setShowModal = () => {},
    fetchPeripherals = () => {},
  } = useContext(PeripheralContext);

  const { handleUpdatePeripheral } = usePeripheralUpdate({
    onSuccess: () => {
      fetchPeripherals();
    },
  });

  const [collected, drop] = useDrop(
    () => ({
      accept: ItemTypes.PERIPHERAL,
      drop: (item: any, monitor: any) => {
        handleUpdatePeripheral({ ...item, gateway: "detach" });
      },
      collect: (monitor) => {
        return monitor.getItem();
      },
    }),
    []
  );

  const unatachedPeripherals = peripherals.filter((p) => !p.gateway);

  return (
    <div ref={drop} style={{ height: "100%" }}>
      <PeripheralTopBar />
      <S.DivH>
        <PeripheralList peripherals={unatachedPeripherals} />
      </S.DivH>

      <S.Modal
        title={"Add Peripheral"}
        maskClosable
        visible={showModal}
        onCancel={() => setShowModal(false)}
        footer={[
          <Button type="primary" onClick={() => setShowModal(false)}>
            Cancel
          </Button>,
        ]}
      >
        <PeripheralCreate />
      </S.Modal>
    </div>
  );
};
