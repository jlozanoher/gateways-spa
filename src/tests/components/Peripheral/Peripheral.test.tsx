import React from "react";
import renderer from "react-test-renderer";
import { Peripheral } from "../../../components/Peripheral/Peripheral";
import { PeripheralModel } from "../../../types/peripheral.model";
import { WithDnDContext } from "../../test_utils";

const peripheral: PeripheralModel = {
  _id: "asd1dasd",
  createdAt: new Date("2022-01-22T10:20:30Z").toISOString(),
  gateway: "123123sdasd1",
  status: "online",
  uid: 123,
  vendor: "Vendor 1",
};

test("Peripheral should match snap", () => {
  const tree = renderer
    .create(
      <WithDnDContext>
        <Peripheral peripheral={peripheral} />
      </WithDnDContext>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
