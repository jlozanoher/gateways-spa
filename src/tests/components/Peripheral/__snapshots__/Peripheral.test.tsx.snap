// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Peripheral should match snap 1`] = `
<div
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <div
    className="ant-card ant-card-bordered ant-card-hoverable sc-pVTFL gQUQji"
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <div
        className="ant-row"
        style={Object {}}
      >
        <div
          className="ant-col ant-col-md-8"
          style={Object {}}
        >
          <button
            className="ant-btn ant-btn-icon-only ant-btn-dangerous"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-label="delete"
              className="anticon anticon-delete"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="delete"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          className="ant-col"
          style={Object {}}
        >
          <div
            className="ant-space ant-space-vertical sc-furwcr mkXde"
            style={Object {}}
          >
            <div
              className="ant-space-item"
              style={
                Object {
                  "marginBottom": 8,
                }
              }
            >
              <span
                className="ant-typography"
                onClick={null}
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                123
              </span>
            </div>
            <div
              className="ant-space-item"
              style={
                Object {
                  "marginBottom": 8,
                }
              }
            >
              <span
                className="ant-typography ant-typography-secondary"
                onClick={null}
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                1231....asd1
              </span>
            </div>
            <div
              className="ant-space-item"
              style={
                Object {
                  "marginBottom": 8,
                }
              }
            >
              <span
                className="ant-typography"
                onClick={null}
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                online
              </span>
            </div>
            <div
              className="ant-space-item"
              style={
                Object {
                  "marginBottom": 8,
                }
              }
            >
              <span
                className="ant-typography"
                onClick={null}
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                Vendor 1
              </span>
            </div>
            <div
              className="ant-space-item"
              style={Object {}}
            >
              <span
                className="ant-typography"
                onClick={null}
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                22-01-22
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
